lib/cli/run-helpers.js

```diff
@@ -0,0 +1,335 @@
+
+/**
+ * Exits Mocha when tests + code under test has finished execution (default)
+ * @param {number} code - Exit code; typically # of failures
+ * @private
+ */
+const exitLater = code => {
+  process.on('exit', () => {
+    process.exit(Math.min(code, 255));
+  });
+};
+
+/**
+ * Exits Mocha when Mocha itself has finished execution, regardless of
+ * what the tests or code under test is doing.
+ * @param {number} code - Exit code; typically # of failures
+ * @private
+ */
+const exit = code => {
+  const clampedCode = Math.min(code, 255);
+  let draining = 0;
+
+  // Eagerly set the process's exit code in case stream.write doesn't
+  // execute its callback before the process terminates.
+  process.exitCode = clampedCode;
+
+  // flush output for Node.js Windows pipe bug
+  // https://github.com/joyent/node/issues/6247 is just one bug example
+  // https://github.com/visionmedia/mocha/issues/333 has a good discussion
+  const done = () => {
+    if (!draining--) {
+      process.exit(clampedCode);
+    }
+  };
+
+  const streams = [process.stdout, process.stderr];
+
+  streams.forEach(stream => {
+    // submit empty write request and wait for completion
+    draining += 1;
+    stream.write('', done);
+  });
+
+  done();
+};
+
+/**
+ * Hide the cursor.
+ * @private
+ */
+const hideCursor = () => {
+  process.stdout.write('\u001b[?25l');
+};
+
+/**
+ * Show the cursor.
+ * @private
+ */
+const showCursor = () => {
+  process.stdout.write('\u001b[?25h');
+};
+
+/**
+ * Stop cursor business
+ * @private
+ */
+const stop = () => {
+  process.stdout.write('\u001b[2K');
+};
+
+/**
+ * Dumps a sorted list of the enumerable, lower-case keys of some object
+ * to `STDOUT`.
+ * @param {Object} obj - Object, ostensibly having some enumerable keys
+ * @private
+ */
+const showKeys = obj => {
+  console.log();
+  Object.keys(obj)
+    .filter(value => /^[a-z]/.test(value))
+    .sort()
+    .forEach(key => {
+      console.log(`    ${key}`);
+    });
+  console.log();
+};
+
+/**
+ * Coerce a comma-delimited string (or array thereof) into a flattened array of
+ * strings
+ * @param {string|string[]} str - Value to coerce
+ * @returns {string[]} Array of strings
+ * @private
+ */
+exports.list = str =>
+  Array.isArray(str) ? exports.list(str.join(',')) : (str || '').split(/ *, */);
+
+/**
+ * Dump list of built-in interfaces
+ * @private
+ */
+exports.runShowInterfaces = () => {
+  showKeys(Mocha.interfaces);
+};
+
+/**
+ * Dump list of built-in reporters
+ * @private
+ */
+exports.runShowReporters = () => {
+  showKeys(Mocha.reporters);
+};
+
+/**
+ * Process extensions for `--watch`, `--compilers`, etc.
+ * @param {yargs.Arguments} argv - Arguments
+ * @private
+ */
+exports.handleExtensions = argv => {
+  const compilers = argv.compilers || [];
+  if (compilers.length) {
+    require('util').deprecate(() => {},
+    '"--compilers" will be removed in a future version of Mocha; see https://git.io/vdcSr for more info')();
+  }
+
```

Comment:
Shouldn't `--compilers` be removed?

Reply:
Any thoughts for an `--extensions` flag to add extensions to the lookup pattern to find tests? Given the retirement of `--compilers`, it would make life easier for CoffeeScript/TypeScript users.

Reply:
`--extensions` is essentially an undocumented flag at this point, so that's probably a good idea.

I have FUD around removing `--compilers`.  But you're right, we should probably do that.

- [ ] expose `--extensions`
- [ ] hard-deprecate `--compilers`

Reply:
@plroebuck I'm unsure how to describe `--extensions`.  Ideas?

Reply:
--extensions <exts> : Add extensions for files to be tested for alternative languages.